src/ChunkGeneratorThread.cpp

```diff
@@ -0,0 +1,256 @@
+#include "Generating/ChunkGenerator.h"
+#include "Generating/ChunkDesc.h"
+
+
+
+
+
+/** A warning is emitted upon adding a chunk to the queue if there are more queued chunks than this. */
+const size_t QUEUE_WARNING_LIMIT = 2000;
+
+/** If there are more than this number chunks in the queue, the generator will skip chunks that have no clients. */
+const size_t QUEUE_SKIP_LIMIT = 1000;
+
+
+
+
+
+cChunkGeneratorThread::cChunkGeneratorThread(void) :
+	Super("cChunkGeneratorThread"),
+	m_Seed(0),  // Will be overwritten by the actual generator
+	m_Generator(nullptr),
+	m_PluginInterface(nullptr),
+	m_ChunkSink(nullptr)
+{
+}
+
+
+
+
+
+cChunkGeneratorThread::~cChunkGeneratorThread()
+{
+	Stop();
+}
+
+
+
+
+
+bool cChunkGeneratorThread::Initialize(cPluginInterface & a_PluginInterface, cChunkSink & a_ChunkSink, cIniFile & a_IniFile)
+{
+	m_PluginInterface = &a_PluginInterface;
+	m_ChunkSink = &a_ChunkSink;
+
+	m_Generator = cChunkGenerator::CreateFromIniFile(a_IniFile);
+	if (m_Generator == nullptr)
+	{
+		LOGERROR("Generator could not start, aborting the server");
+		return false;
+	}
+	return true;
+}
+
+
+
+
+
+void cChunkGeneratorThread::Stop(void)
+{
+	m_ShouldTerminate = true;
+	m_Event.Set();
+	m_evtRemoved.Set();  // Wake up anybody waiting for empty queue
+	Super::Stop();
+	m_Generator.reset();
+}
+
+
+
+
+
+void cChunkGeneratorThread::QueueGenerateChunk(
+	cChunkCoords a_Coords,
+	bool a_ForceRegeneration,
+	cChunkCoordCallback * a_Callback
+)
+{
+	ASSERT(m_ChunkSink->IsChunkQueued(a_Coords));
+
+	{
+		cCSLock Lock(m_CS);
+
+		// Add to queue, issue a warning if too many:
+		if (m_Queue.size() >= QUEUE_WARNING_LIMIT)
+		{
+			LOGWARN("WARNING: Adding chunk %s to generation queue; Queue is too big! (%zu)", a_Coords.ToString().c_str(), m_Queue.size());
+		}
+		m_Queue.push_back(QueueItem{a_Coords, a_ForceRegeneration, a_Callback});
+	}
+
+	m_Event.Set();
+}
+
+
+
+
+
+void cChunkGeneratorThread::GenerateBiomes(cChunkCoords a_Coords, cChunkDef::BiomeMap & a_BiomeMap)
+{
+	if (m_Generator != nullptr)
+	{
+		m_Generator->GenerateBiomes(a_Coords.m_ChunkX, a_Coords.m_ChunkZ, a_BiomeMap);
+	}
+}
+
+
+
+
+
+void cChunkGeneratorThread::WaitForQueueEmpty(void)
+{
+	cCSLock Lock(m_CS);
+	while (!m_ShouldTerminate && !m_Queue.empty())
+	{
+		cCSUnlock Unlock(Lock);
+		m_evtRemoved.Wait();
+	}
+}
+
+
+
+
+
+int cChunkGeneratorThread::GetQueueLength(void)
+{
+	cCSLock Lock(m_CS);
+	return static_cast<int>(m_Queue.size());
+}
+
+
+
+
+
+EMCSBiome cChunkGeneratorThread::GetBiomeAt(int a_BlockX, int a_BlockZ)
+{
+	ASSERT(m_Generator != nullptr);
+	return m_Generator->GetBiomeAt(a_BlockX, a_BlockZ);
+}
+
+
+
+
+
+void cChunkGeneratorThread::Execute(void)
+{
+	// To be able to display performance information, the generator counts the chunks generated.
+	// When the queue gets empty, the count is reset, so that waiting for the queue is not counted into the total time.
+	int NumChunksGenerated = 0;  // Number of chunks generated since the queue was last empty
```

Comment:
Shouldn't variables start lower-cased?

Reply:
This is basically only code that has been moved, so I didn't change it.
Subconsciously I don't use the new conventions on old code just yet, I'll change everything at once when changing the API.